### PR TITLE
AppendSpectra now properly sets the values of TextAxis and NumericAxis

### LINKED ISSUE #17902: AppendSpectra ignores the values of TextAxis and Numeric, producing output workspace with empty axis labels
### Expected behavior

After appending workspaces using AppendSpectra with TextAxis and Numeric, the axis labels are kept for each spectra from each workspace.
### Actual behavior

The axis labels are discarded and the produced workspace has no labels unless the axis contained spectrum numbers.
### Steps to reproduce the behavior
- Append 2 workspaces that have TextAxis or Numeric axis
- Output workspace will have the same number of spectra as the input workspaces, but the labels will be missing.
### Platforms affected

Tested only on Windows 7, but probably all
#### Relevant Docs

[TextAxis docs](http://docs.mantidproject.org/nightly/api/python/mantid/api/TextAxis.html)
[MantidAxis docs](http://docs.mantidproject.org/nightly/api/python/mantid/api/MantidAxis.html)
[Algorithm reference, AppendSpectra.cpp L119](https://github.com/mantidproject/mantid/blob/master/Framework/Algorithms/src/AppendSpectra.cpp#L119)

### PR DESCRIPTION
**Pull request by dtasev**
Description of work.

`AppendSpectra` algorithm used to ignore the values of the TextAxis and NumericAxis labels and just append the spectra number. This is fixed in this PR and the algorithm now properly appends the value of TextAxis and NumericAxis, while keeping the old behaviour unchanged.

**To test:**
- I've made a quite long testing script, that I used to test all cases I could think of, it can be found here: https://gist.github.com/DimitarTasev/1c16fbfe60b95f8dff4845fbb7b02393
- Run unit tests and make sure they pass successfully
- Code review, changes aren't that big, but if you can think of any better way to handle the unique cases, please tell me

Fixes #17902 .

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/AppendSpectra.cpp [hunk 1]
```diff
@@ -131,10 +133,46 @@ void AppendSpectra::fixSpectrumNumbers(API::MatrixWorkspace_const_sptr ws1,
...
+      output->getSpectrum(i).setSpectrumNo(specnum_t(i));
+    }
+
+  } else {
+    // change the axis by adding the maximum existing spectrum number to the
+    // current value
+    // default to old behaviour if none of our special cases are true
+    for (size_t i = 0; i < output->getNumberHistograms(); i++) {
+      output->getSpectrum(i).setSpectrumNo(specnum_t(i));
+    }
```
Comment (soininen) [inline Framework/Algorithms/src/AppendSpectra.cpp]: If I understood the above code right, you could also consider writing this if-else if-else thing the other way around:
```C++
  for (size_t i = 0; i < output->getNumberHistograms(); ++i) {
    if (isTextAxis) {
      const std::string inputLabel = i < ws1len ? yAxisWS1->label(i) : yAxisWS2->label(i - ws1len);
      outputTextAxis->setLabel(i, (inputLabel.size() > 0) ? inputLabel : "");
    }
    else if (isNumericAxis) {
      const double inputVal = i < ws1len ? yAxisWS1->getValue(i) : yAxisWS2->getValue(i - ws1len);
      outputYAxis->setValue(i, inputVal);
    }
    output->getSpectrum(i).setSpectrumNo(specnum_t(i));
  }
```
Less code duplication, no?

File: Framework/Algorithms/test/AppendSpectraTest.h [hunk 2]
```diff
@@ -25,210 +27,429 @@ class AppendSpectraTest : public CxxTest::TestSuite {
...
-
-    for (size_t wi = 0; wi < out->getNumberHistograms(); wi++) {
-      TS_ASSERT_EQUALS(out->getSpectrum(wi).getSpectrumNo(), specnum_t(wi));
-      TS_ASSERT(!out->getSpectrum(wi).getDetectorIDs().empty());
-      for (size_t x = 0; x < out->blocksize(); x++)
-        TS_ASSERT_DELTA(out->readY(wi)[x], 2.0, 1e-5);
+        appendSpectra->setProperty("InputWorkspace2", inputWorkspace2));
+    TS_ASSERT_THROWS_NOTHING(
+        appendSpectra->setProperty("OutputWorkspace", outputWorkspace));
+    TS_ASSERT_THROWS_NOTHING(appendSpectra->execute());
```
Comment (soininen) [inline Framework/Algorithms/test/AppendSpectraTest.h]: To my understanding, call to execute() never throws, unless you do `appendSpectra->setRethrows(true);`.

**Review by soininen — CHANGES_REQUESTED** [review]

**Comment by soininen** [comment]
Is this also an issue in [`ConjoinWorkspaces`](http://docs.mantidproject.org/nightly/algorithms/ConjoinWorkspaces-v1.html)?

**Comment by DimitarTasev** [comment]
@soininen Yes this is also an issue in [`ConjoinWorkspaces`](http://docs.mantidproject.org/nightly/algorithms/ConjoinWorkspaces-v1.html), the text axis was dropped

**Comment by DimitarTasev** [comment]
Restarted [RHEL 10042](http://builds.mantidproject.org/job/pull_requests-rhel6/10042/) due to unrelated failure

**Review by soininen — APPROVED** [review]

**Comment by soininen** [comment]
Everything seems to be in order. :shipit:

**Pull request merged**